src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
```

Comment:
Nit: Typically the idiomatic naming scheme for `enum` types doesn't include a pluralization. So, idiomatic rust would call this:
```suggestion
pub enum DecodeError {
```
You may want to change the names of `ApiKeys` and `ApiVersions` as well, though I recognize that those are the names of the types in the Kafka specification.